rueidisprob/countingbloomfilter.go

```diff
@@ -0,0 +1,369 @@
+	Count(ctx context.Context) (uint, error)
+}
+
+type countingBloomFilter struct {
+	client rueidis.Client
+
+	// name is the name of the Counting Bloom Filter.
+	// It is used as a key in the Redis.
+	name string
+
+	// counter is the name of the counter.
+	counter string
+
+	// hashIterations is the number of hash functions to use.
+	hashIterations      uint
+	hashIterationString string
+
+	// size is the number of bits to use.
+	size uint
+
+	addMultiScript *rueidis.Lua
+	addMultiKeys   []string
+
+	existsMultiScript *rueidis.Lua
+	existsMultiKeys   []string
+
+	removeMultiScript *rueidis.Lua
+	removeMultiKeys   []string
+}
+
+// NewCountingBloomFilter creates a new Counting Bloom Filter.
+// NOTE: 'name:cbf:c' is used as a counter key in the Redis and
+// 'name:cbf' is used as a filter key in the Redis
+// to keep track of the number of items in the Counting Bloom Filter for Count method.
+func NewCountingBloomFilter(
+	client rueidis.Client,
+	name string,
+	expectedNumberOfItems uint,
+	falsePositiveRate float64,
+) (CountingBloomFilter, error) {
+	if len(name) == 0 {
+		return nil, ErrEmptyCountingBloomFilterName
+	}
+
+	if falsePositiveRate <= 0 {
+		return nil, ErrCountingBloomFilterFalsePositiveRateLessThanEqualZero
+	}
+	if falsePositiveRate >= 1 {
+		return nil, ErrCountingBloomFilterFalsePositiveRateGreaterThanOne
+	}
+
+	size := numberOfBloomFilterBits(expectedNumberOfItems, falsePositiveRate)
+	if size == 0 {
+		return nil, ErrCountingBloomFilterBitsSizeZero
+	}
+	hashIterations := numberOfBloomFilterHashFunctions(size, expectedNumberOfItems)
+
+	// NOTE: https://redis.io/docs/reference/cluster-spec/#hash-tags
+	baseName := "{" + name + "}"
+	bfName := baseName + ":cbf"
+	counterName := bfName + ":c"
+	return &countingBloomFilter{
+		client:              client,
+		name:                bfName,
+		counter:             counterName,
+		hashIterations:      hashIterations,
+		hashIterationString: strconv.FormatUint(uint64(hashIterations), 10),
+		size:                size,
+		addMultiScript:      rueidis.NewLuaScript(countingBloomFilterAddMultiScript),
+		addMultiKeys:        []string{bfName, counterName},
+		existsMultiScript:   rueidis.NewLuaScript(countingBloomFilterExistsMultiScript),
+		existsMultiKeys:     []string{bfName},
+		removeMultiScript:   rueidis.NewLuaScript(countingBloomFilterRemoveMultiScript),
+		removeMultiKeys:     []string{bfName, counterName},
+	}, nil
+}
+
+func (f *countingBloomFilter) Add(ctx context.Context, key string) error {
+	return f.AddMulti(ctx, []string{key})
+}
+
+func (f *countingBloomFilter) AddMulti(ctx context.Context, keys []string) error {
+	if len(keys) == 0 {
+		return nil
+	}
+
+	indexes := f.indexes(keys)
+
+	args := make([]string, 0, len(indexes)+1)
+	args = append(args, strconv.Itoa(len(keys)))
+	args = append(args, indexes...)
+
+	resp := f.addMultiScript.Exec(ctx, f.client, f.addMultiKeys, args)
+	if resp.Error() != nil {
+		return resp.Error()
+	}
+
+	return nil
```

Comment:
```suggestion
	return resp.Error()
```
I am not sure if this kind of pattern can be optimized by the go compiler but we can just shortcut it by hand.

Reply:
[bc2f494](https://github.com/redis/rueidis/pull/519/commits/bc2f494e049ece15d4fd12635c18d55c207c7138)